### ISSUE TITLE
IOperation API shape for object and collection initializer expressions

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn#18781]
Currently, BoundCollectionIntializerExpression has OperationKind.None and implements IOperation. This means that we completely cut-off the operations within a collection initializer, including even the nested initializers.

We need to design the API shape for Collection initializers.

1. Do we need an `ICollectionInitializerExpression`? Theoretically, we don't need it as `IObjectCreationExpression.Initializers` can return the child collection element initializer nodes of a collection initializer.
2. Do we need an `ICollectionElementInitializerExpression`? We have 2 possible designs:
   1. Follow the VB bound tree pattern, which stores a `BoundCall` for collection element initializer, so we don't need a new API. However, C# doesn't store this bound call, but instead has a `BoundCollectionElementInitializerExpression` (with invoked method name and arguments) and `BoundDynamicCollectionElementInitializer` (with possible applicable methods and arguments) . So we need to synthesize such a bound call for C# and return it as an initializer. Note that this approach introduces an implicit compiler generated IOperation node in the operation tree.
   2. Add the following API and our Operation tree node for CollectionElementInitializerExpression, can have language specific implementations that delegate to either BoundCall (VB) or BoundICollectionElementInitializerExpression or BoundDynamicCollectionElementInitializer (C#):
```
namespace Microsoft.CodeAnalysis.Semantics
{
    /// <summary>
    /// Represents a collection element initializer expression.
    /// </summary>
    /// <remarks>
    /// This interface is reserved for implementation by its associated APIs. We reserve the right to
    /// change it in the future.
    /// </remarks>
    public interface ICollectionElementInitializerExpression : IHasArgumentsExpression
    {
        /// <summary>
        /// Set of applicable methods for an implicit dynamic invocation OR the implicit Add method symbol for non-dynamic invocation.
        /// </summary>
        ImmutableArray<IMethodSymbol> ApplicableMethods { get; }

        /// <summary>
        /// Flag indicating if this is a dynamic initializer invocation.
        /// </summary>
        bool IsDynamic { get; }
    }
}
```

look for an example here  https://github.com/dotnet/roslyn/issues/18509

**Comment by jinujoseph** [comment]
also look into https://github.com/dotnet/roslyn/issues/8883

**Comment by mavasani** [comment]
API proposal:

```
/// <summary>
    /// Represents a collection element initializer expression.
    /// </summary>
    /// <remarks>
    /// This interface is reserved for implementation by its associated APIs. We reserve the right to
    /// change it in the future.
    /// </remarks>
    public interface ICollectionElementInitializerExpression
    {
        /// <summary>
        /// Initializers
        /// </summary>
        ImmutableArray<IOperation> Initializers { get; }

        /// <summary>
        /// Set of applicable methods for an implicit dynamic invocation OR the implicit Add method symbol for non-dynamic invocation.
        /// </summary>
        ImmutableArray<IMethodSymbol> ApplicableMethods { get; }

        /// <summary>
        /// Flag indicating if this is a dynamic initializer invocation.
        /// </summary>
        bool IsDynamic { get; }
    }
```

Examples:

1. Collection element initializer: http://source.roslyn.io/#Roslyn.Compilers.CSharp.Semantic.UnitTests/IOperation/IOperationTests_IParameterReferenceExpression.cs,173
2. Dynamic collection element initializer: http://source.roslyn.io/#Roslyn.Compilers.CSharp.Semantic.UnitTests/IOperation/IOperationTests_IParameterReferenceExpression.cs,308

**Comment by CyrusNajmabadi** [comment]
Replace ApplicableMethod with just a single IMethodSymbol.  Can be null in dynamic case.  Keep IsDynamic.

Initializers  should probably be Arguments.

**Comment by CyrusNajmabadi** [comment]
For something like ```X = x, Y = { x, y, 3 }```   we do want an assignment initializer for "X = x" (because the field/prop really is being assigned).  However, for "Y = ..." we want something new.  Something that says "this is actually a collection that we're going to .Add on for all these elements".  

For "C = { X = z }"   there is an assignment to "X", but not to "C".  It's an assignment *through* C.

**Comment by mavasani** [comment]
@AlekseyTs and I discussed this further and came up with the following API proposal:

```
namespace Microsoft.CodeAnalysis.Semantics
{
    /// <summary>
    /// Represents a C# or VB object or collection initializer expression.
    /// For example, object initializer "{ X = x }" within object creation "new Class() { X = x }" and
    /// collection initializer "{ x, y, 3 }" within collection creation "new MyList() { x, y, 3 }".
    /// </summary>
    /// <remarks>
    /// This interface is reserved for implementation by its associated APIs. We reserve the right to
    /// change it in the future.
    /// </remarks>
    public interface IObjectOrCollectionInitializerExpression : IOperation
    {
        /// <summary>
        /// Object member or collection initializers.
        /// </summary>
        ImmutableArray<IOperation> Initializers { get; }
    }

    /// <summary>
    /// Represents a C# or VB member initializer expression within an object initializer expression.
    /// For example, given an object creation with initializer "new Class() { X = x, Y = { x, y, 3 }, Z = { X = z } }",
    /// member initializers for Y and Z, i.e. "Y = { x, y, 3 }", and "Z = { X = z }" are nested member initializers represented by this operation.
    /// </summary>
    /// <remarks>
    /// This interface is reserved for implementation by its associated APIs. We reserve the right to
    /// change it in the future.
    /// </remarks>
    public interface IMemberInitializerExpression : IOperation
    {
        /// <summary>
        /// Initialized member.
        /// </summary>
        IMemberReferenceExpression InitializedMember { get; }
        
        /// <summary>
        /// Member initializer.
        /// </summary>
        IObjectOrCollectionInitializerExpression Initializer { get; }
    }

    /// <summary>
    /// Represents a C# nested collection element initializer expression within a collection initializer.
    /// For example, given a collection initializer "new Class() { Y = { { x, y, 3 } } }",
    /// nested collection element initializer for Y, i.e. "{ { x, y, 3 } }" is represented by this operation.
    /// </summary>
    /// <remarks>
    /// This interface is reserved for implementation by its associated APIs. We reserve the right to
    /// change it in the future.
    /// </remarks>
    public interface ICollectionElementInitializerExpression : IOperation
    {
        /// <summary>
        /// Add method invoked on collection. Might be null for dynamic invocation.
        /// </summary>
        IMethodSymbol AddMethod { get; }

        /// <summary>
        /// Arguments passed to add method invocation.
        /// </summary>
        ImmutableArray<IOperation> Arguments { get; }

        /// <summary>
        /// Flag indicating if this is a dynamic invocation.
        /// </summary>
        bool IsDynamic { get; }
    }
}
```

The following example with provide more clarity:
**Source code:**
```
using System.Collections.Generic;

internal class Class
{
    public int X { get; set; }
    public List<int> Y { get; set; }
    public Dictionary<int, int> Z { get; set; }
    public Class C { get; set; }

    public void M(int x, int y, int z)
    {
        var c = /*<bind>*/new Class() { X = x, Y = { x, y, 3 }, Z = { { x, y } }, C = { X = z } }/*</bind>*/;

        var d = new List<int> { x, y };
    }
}
```

**Operation tree:**
```
IOperation tree for "new Class() { X = x, Y = { x, y, 3 }, Z = { { x, y } }, C = { X = z } }"

IObjectCreationExpression (Constructor: Class..ctor()) (OperationKind.ObjectCreationExpression, Type: Class) (Syntax: 'new Class() ... { X = z } }')
  Initializer: IObjectOrCollectionInitializerExpression (OperationKind.ObjectOrCollectionInitializerExpression, Type: Class) (Syntax: '{ X = x, Y  ... { X = z } }')
      Initializers(4): IAssignmentExpression (OperationKind.AssignmentExpression, Type: System.Int32) (Syntax: 'X = x')
          Left: IPropertyReferenceExpression: System.Int32 Class.X { get; set; } (Static) (OperationKind.PropertyReferenceExpression, Type: System.Int32) (Syntax: 'X')
          Right: IParameterReferenceExpression: x (OperationKind.ParameterReferenceExpression, Type: System.Int32) (Syntax: 'x')
        IMemberInitializerExpression (OperationKind.MemberInitializerExpression, Type: System.Collections.Generic.List<System.Int32>) (Syntax: 'Y = { x, y, 3 }')
          InitializedMember: IPropertyReferenceExpression: System.Collections.Generic.List<System.Int32> Class.Y { get; set; } (Static) (OperationKind.PropertyReferenceExpression, Type: System.Collections.Generic.List<System.Int32>) (Syntax: 'Y')
          Initializer: IObjectOrCollectionInitializerExpression (OperationKind.ObjectOrCollectionInitializerExpression, Type: System.Collections.Generic.List<System.Int32>) (Syntax: '{ x, y, 3 }')
              Initializers(3): ICollectionElementInitializerExpression (AddMethod: void System.Collections.Generic.List<System.Int32>.Add(System.Int32 item)) (IsDynamic: False) (OperationKind.CollectionElementInitializerExpression, Type: System.Void) (Syntax: 'x')
                  Arguments(1): IParameterReferenceExpression: x (OperationKind.ParameterReferenceExpression, Type: System.Int32) (Syntax: 'x')
                ICollectionElementInitializerExpression (AddMethod: void System.Collections.Generic.List<System.Int32>.Add(System.Int32 item)) (IsDynamic: False) (OperationKind.CollectionElementInitializerExpression, Type: System.Void) (Syntax: 'y')
                  Arguments(1): IParameterReferenceExpression: y (OperationKind.ParameterReferenceExpression, Type: System.Int32) (Syntax: 'y')
                ICollectionElementInitializerExpression (AddMethod: void System.Collections.Generic.List<System.Int32>.Add(System.Int32 item)) (IsDynamic: False) (OperationKind.CollectionElementInitializerExpression, Type: System.Void) (Syntax: '3')
                  Arguments(1): ILiteralExpression (Text: 3) (OperationKind.LiteralExpression, Type: System.Int32, Constant: 3) (Syntax: '3')
        IMemberInitializerExpression (OperationKind.MemberInitializerExpression, Type: System.Collections.Generic.Dictionary<System.Int32, System.Int32>) (Syntax: 'Z = { { x, y } }')
          InitializedMember: IPropertyReferenceExpression: System.Collections.Generic.Dictionary<System.Int32, System.Int32> Class.Z { get; set; } (Static) (OperationKind.PropertyReferenceExpression, Type: System.Collections.Generic.Dictionary<System.Int32, System.Int32>) (Syntax: 'Z')
          Initializer: IObjectOrCollectionInitializerExpression (OperationKind.ObjectOrCollectionInitializerExpression, Type: System.Collections.Generic.Dictionary<System.Int32, System.Int32>) (Syntax: '{ { x, y } }')
              Initializers(1): ICollectionElementInitializerExpression (AddMethod: void System.Collections.Generic.Dictionary<System.Int32, System.Int32>.Add(System.Int32 key, System.Int32 value)) (IsDynamic: False) (OperationKind.CollectionElementInitializerExpression, Type: System.Void) (Syntax: '{ x, y }')
                  Arguments(2): IParameterReferenceExpression: x (OperationKind.ParameterReferenceExpression, Type: System.Int32) (Syntax: 'x')
                    IParameterReferenceExpression: y (OperationKind.ParameterReferenceExpression, Type: System.Int32) (Syntax: 'y')
        IMemberInitializerExpression (OperationKind.MemberInitializerExpression, Type: Class) (Syntax: 'C = { X = z }')
          InitializedMember: IPropertyReferenceExpression: Class Class.C { get; set; } (Static) (OperationKind.PropertyReferenceExpression, Type: Class) (Syntax: 'C')
          Initializer: IObjectOrCollectionInitializerExpression (OperationKind.ObjectOrCollectionInitializerExpression, Type: Class) (Syntax: '{ X = z }')
              Initializers(1): IAssignmentExpression (OperationKind.AssignmentExpression, Type: System.Int32) (Syntax: 'X = z')
                  Left: IPropertyReferenceExpression: System.Int32 Class.X { get; set; } (Static) (OperationKind.PropertyReferenceExpression, Type: System.Int32) (Syntax: 'X')
                  Right: IParameterReferenceExpression: z (OperationKind.ParameterReferenceExpression, Type: System.Int32) (Syntax: 'z')
```

**Comment by mavasani** [comment]
Fixed with #20689
